python/search_widget/search_widget.py

```diff
@@ -0,0 +1,150 @@
+        
+        # set up the UI
+        self._ui = Ui_SearchWidget()
+        self._ui.setupUi(self)
+        
+        # dynamically create the clear button so that we can place it over the
+        # edit widget:
+        self._clear_btn = QtGui.QPushButton(self._ui.search_edit)
+        self._clear_btn.setFocusPolicy(QtCore.Qt.StrongFocus)
+        self._clear_btn.setFlat(True)
+        self._clear_btn.setCursor(QtCore.Qt.ArrowCursor)
+        style = ("QPushButton {"
+                 + "border: 0px solid;"
+                 + "image: url(:/tk-framework-qtwidgets/search_widget/clear_search.png);"
+                 + "width: 16;"
+                 + "height: 16;"
+                 + "}"
+                 + "QPushButton::hover {"
+                 + "image: url(:/tk-framework-qtwidgets/search_widget/clear_search_hover.png);"
+                 + "}")
+        self._clear_btn.setStyleSheet(style)
+        self._clear_btn.hide()
+        
+        h_layout = QtGui.QHBoxLayout(self._ui.search_edit)
+        h_layout.addStretch()
+        h_layout.addWidget(self._clear_btn)
+        h_layout.setContentsMargins(3, 0, 3, 0)
+        h_layout.setSpacing(0)
+        self._ui.search_edit.setLayout(h_layout)
+
+        # hook up the signals:
+        self._ui.search_edit.textEdited.connect(self._on_text_edited)
+        self._ui.search_edit.returnPressed.connect(self._on_return_pressed)
+        self._clear_btn.clicked.connect(self._on_clear_clicked)
+                
+    # @property
+    def _get_search_text(self):
+        """
+        get the search text from the widget
+        """
+        text = self._ui.search_edit.text()
+        return self._safe_to_string(text)
+
+    # @search_text.setter
+    def _set_search_text(self, value):
+        """
+        set the search text on the widget
+        """
+        self._ui.search_edit.setText(value)
+    search_text = property(_get_search_text, _set_search_text)
+
+    def set_placeholder_text(self, text):
+        """
+        Set the placeholder text for the widget
+        
+        :param text:    The text to use
+        """
+        self._ui.search_edit.setPlaceholderText(text)
+                
+    def _on_clear_clicked(self):
+        """
+        Slot triggered when the clear button is clicked - clears the text
+        and emits the relevant signals.
+        """
+        self._ui.search_edit.setText("")
+        self.search_changed.emit("")
+        self.search_edited.emit("")
+        self._clear_btn.hide()
+                
+    def _on_text_edited(self):
+        """
+        Slot triggered when the text has been edited 
+        """
+        text = self.search_text
+        self._clear_btn.setVisible(bool(text))
+        self.search_edited.emit(text)
+        
+    def _on_return_pressed(self):
+        """
+        Slot triggered when return has been pressed
+        """
+        self.search_changed.emit(self.search_text)
+        
+    def _safe_to_string(self, value):
```

Comment:
This looks like a utility that should be present in core or one of the frameworks.